README.md
# ScanByte.jl

![CI](https://github.com/jakobnissen/ScanByte.jl/workflows/CI/badge.svg)
[![Codecov](https://codecov.io/gh/jakobnissen/ScanByte.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/jakobnissen/ScanByte.jl)

_Find your bytes. Fast._

ScanByte is a package to solve a simple problem: Find the first occurrence of a byte or any byte from a set of bytes in a chunk of memory. Think of it like a much faster version of `findfirst` that only iterates over bytes in memory.

ScanByte is micro-optimized for speed. On my laptop it can hit the RAM bandwidth limit of around 20 GB/s. This speed makes it a suitable building block for string search engines, Regex implementations, parsers and similar use cases.

### Usage
The central function of interest in this package is `memchr`.
The function takes a chunk of memory and some bytes, and returns the first position (1-indexed) in the chunk of memory where any byte from the byte set is found, or `nothing` if no bytes are found.

The chunk of memory can be any type which implements methods for `pointer` and `sizeof`, or alternatively you can input a raw pointer and a size.

The byte set can be passed in in two different ways:
* As a single `UInt8`, in which case ScanByte will simply dispatch to libc's memchr function
* A `Val{bs}`, where `bs` is an instance of the type `ByteSet <: AbstractSet{UInt8}` from this package.

In the latter case, ScanByte will, at compile time, pick an efficient SIMD algorithm based on the content of the byteset.
Currently ScanByte only has SIMD algorithms for SSSE3 and AVX2 instruction sets (found in all x86-based PCs), and uses a slow fallback for CPUs without these instructions.

### Example usage
In all these examples, the input data can be a `String`, a `codeunits` object, an `Array{UInt8}`, or a pointer+length.
Any type that implements `pointer` and `sizeof` will do.

Search for a single byte:
```julia
julia> memchr("Julia", UInt8('i'))
4

julia> memchr(codeunits("Julia"), UInt8('z')) === nothing
true

julia> str = "Julia";

julia> GC.@preserve str memchr(pointer(str), sizeof(str) % UInt, UInt8('i'))
4
```

Search for a byteset. Here, `Val` must be used to force specialization on the byteset:
```julia
julia> bs = ByteSet([0x01, 0x6a, 0xf1]);

julia> memchr([0x4a, 0xf1], Val(bs))
3
```

Search using a function. To do this, you must construct a `ByteSet` using the predicate on `0x00:0xff`:
```julia
julia> f(x) = in(x, 0x1a:0x4c) || in(x, 0xd1:0xf1); # some function

julia> bs = ByteSet(filter(f, 0x00:0xff));

julia> memchr("hello, Bob", Val(bs))
6
```

## Drawbacks
At the moment, ScanByte has three major drawbacks:

* If you are search for a predicate/byteset and not a single byte, tt relies on generated functions to compute the optimal Julia code to create the scanning function. This means the byte set must be known at compile time.
* It relies on explicit SIMD instructions. To be fast, it must run on computers with ideally the `AVX2` instruction set, or with the `SSE2` and `SSSE3` sets. Also, if you create the scanning function on a computer with `AVX2` but runs it on a computer without, LLVM will probably crash. Currently, the fallback methods are fairly slow.
* There is no guaranteed stable version of detecting which SIMD instructions your Julia supports. So this package tries to guess by parsing some output from LLVM.


NEWS.md
## Release 0.4.0
__Breaking changes__

* Breaking: `SizedMemory` is now internal

__Other_
_
* Now falls back to generic code if `ENV["JULIA_CPU_TARGET"]` is set.
* Use libc memchr when byteset length is one
* Add more set ops to `ByteSet`

## Release 0.3.2
* Optimise ByteSet
* Optimise `memchr(x, ::Byte)` by calling libc's memchr
* Add generic method `memchr(::Any, x)`, automatically creating `SizedMemory` from x

## Release 0.3.1
Fix issue #3 - SIMD fallback warning only emitted at package precompile time.

## Release 0.3
__Breaking changes__

* `gen_scan_function` has been removed. Instead, you should now scan for multiple bytes by calling `memchr(::SizedMemory, Val(byteset))`.

## Release 0.2
__Breaking changes__

* `SizedMemory(x)` now requires that `x` implements `sizeof`

__New features__

* New function `memchr(x::SizedMemory, b::UInt8)`, which finds the first position of `b` in `x`.

## Release 0.1
Initial release


test/byteset.jl
function test_membership(members)
    bs = ByteSet(members)
    refset = Set{UInt8}([UInt8(i) for i in  members])
    @test refset == Set{UInt8}(collect(bs))
    @test all(i -> in(i, bs), refset)
end

function test_inversion(bs)
    inv = ~bs
    all = true
    for i in 0x00:0xff
        all &= (in(i, bs) ⊻ in(i, inv))
    end
    @test all
end

@testset "Instantiation" begin
    @test isempty(ByteSet())
    @test iszero(length(ByteSet()))

    for set in ["hello", "kdjy82zxxcbnpw", [0x00, 0x1a, 0xff, 0xf8, 0xd2]]
        test_membership(set)
    end
end

@testset "Min/max" begin
    @test_throws ArgumentError maximum(ByteSet())
    @test_throws ArgumentError minimum(ByteSet())
    @test minimum(ByteSet("xylophone")) == UInt8('e')
    @test maximum(ByteSet([0xa1, 0x0f, 0x4e, 0xf1, 0x40, 0x39])) == 0xf1
end

@testset "Contiguity" begin
    @test ScanByte.is_contiguous(ByteSet(0x03:0x41))
    @test ScanByte.is_contiguous(ByteSet())
    @test ScanByte.is_contiguous(ByteSet(0x51))
    @test ScanByte.is_contiguous(ByteSet(0xc1:0xd2))
    @test ScanByte.is_contiguous(ByteSet(0x00:0xff))

    @test !ScanByte.is_contiguous(ByteSet([0x12:0x3a; 0x3c:0x4a]))
    @test !ScanByte.is_contiguous(ByteSet([0x01, 0x02, 0x04, 0x05]))
end

@testset "Inversion" begin
    test_inversion(ByteSet())
    test_inversion(ByteSet(0x00:0xff))
    test_inversion(ByteSet([0x04, 0x06, 0x91, 0x92]))
    test_inversion(ByteSet(0x54:0x71))
    test_inversion(ByteSet(0x12:0x11))
    test_inversion(ByteSet("abracadabra"))
end

@testset "Set operations" begin
    sets = map(ByteSet, [
        [],
        [0x00:0xff;],
        [0x00:0x02; 0x04; 0x19],
        [0x01; 0x03; 0x09; 0xa1; 0xa1],
        [0x41:0x8f; 0xd1:0xe1; 0xa0:0xf0],
        [0x81:0x89; 0xd0:0xd0]
    ])
    ssets = map(Set, sets)
    for (s1, ss1) in zip(sets, ssets), (s2, ss2) in zip(sets, ssets)
        for f in [union, intersect, symdiff, setdiff]
            @test Set(f(s1, s2)) == f(ss1, ss2)
        end
        @test isdisjoint(s1, s2) == isdisjoint(ss1, ss2)
    end
end


test/runtests.jl
using ScanByte
using Test

include("byteset.jl")
include("scanbyte.jl")

test/scanbyte.jl
SizedMemory = ScanByte.SizedMemory

@testset "SizedMemory" begin
    for str in ["Hello, world!", "", "αβγ", "    "]
        mem = SizedMemory(str)
        @test pointer(mem) == pointer(str)
        @test length(mem) == ncodeunits(str)

        arr = collect(codeunits(str))
        mem = SizedMemory(arr)
        @test pointer(mem) == pointer(arr)
        @test length(mem) == length(arr)
    end

    for str in ["Hello, world!", "αβγδ", "          "]
        for start in (1, 3), len in (0, 3, 5)
            substr = SubString(str, start:start + len - 1)
            mem = SizedMemory(substr)
            @test pointer(mem) == pointer(substr)
            @test length(mem) == ncodeunits(substr)
        end
    end
end

@testset "memchr of byte" begin
    @test memchr(SizedMemory([]), 0x01) === nothing
    @test memchr(SizedMemory([1,2,3]), 0x02) === 9
    @test memchr(SizedMemory([1,2,3]), 0x04) === nothing
    
    for byte in [0x00, 0xa0, 0xda, 0xff]
        @test memchr(SizedMemory(UInt8[]), byte) === nothing
        @test memchr(UInt8[], byte) === nothing
        not_bytes = [i for i in 0x00:0xff if i != byte]
        for (len, ind) in [(2, 2), (20, 14), (500, 431)]
            v = rand(not_bytes, len)
            @test memchr(SizedMemory(v), byte) === nothing
            @test memchr(v, byte) === nothing
            v[ind] = byte
            @test memchr(SizedMemory(v), byte) == ind
            @test memchr(v, byte) == ind
        end
    end

    # Test the memchr method with pointer
    bytes = [1,2,3]
    @test memchr(pointer(bytes), UInt(11), 0x02) === 9
    @test memchr(pointer(bytes), UInt(11), 0x03) === nothing
end
            

@testset "Scanning" begin
    # We test the inverted bytesets, because the codegen is built to look for
    # the bytes its NOT scanning for. Just to make the test easier
    for T in (ScanByte.v128, ScanByte.v256, nothing)
        for inv_byteset in [
            ByteSet(), # empty
            ByteSet(0x00:0xff), # full
            ByteSet(0x41), # one member
            ~ByteSet(0x6a), # all but one member
            ByteSet([0xa2, 0xb4, 0xc8, 0xf9, 0xf6]), # unique lower nibble
            ~ByteSet([0x38, 0x40, 0x51, 0x79, 0x94]), # inv uniq nibble
            ByteSet(0x49:0x71), # Contiguous
            ~ByteSet(0x02:0x31), # Inverted contiguous
            ByteSet(rand(0x80:0xff, 20)), # 128 - 255
            ByteSet(rand(0x00:0x7f, 20)), # 0 - 127
            ByteSet(rand(0x5a:0x94, 20)), # within 127 of each other
            ~ByteSet(rand(0x5a:0x94, 20)), # inv within 127
            ByteSet([0x38, 0x40, 0x90, 0xba, 0xc5, 0xc8, 0xe7]), # at most 8 elements
            ByteSet([i for i in 0x00:0xff if rand(Bool)]) # fallback
        ]
            byteset = ~inv_byteset

            # Empty
            @test ScanByte._memchr(T, SizedMemory(UInt8[]), Val(byteset)) === nothing

            # 1000 bytes not in the set
            if length(byteset) != 256
                bytes = rand(collect(~byteset), 1000)
                @test ScanByte._memchr(T, SizedMemory(bytes), Val(byteset)) === nothing

                if !isempty(byteset)
                    bytes[500] = first(iterate(byteset))
                    @test ScanByte._memchr(T, SizedMemory(bytes), Val(byteset)) == 500

                    bytes[25] = first(iterate(byteset))
                    @test ScanByte._memchr(T, SizedMemory(bytes), Val(byteset)) == 25
                end

                resize!(bytes, 20)
                @test ScanByte._memchr(T, SizedMemory(bytes), Val(byteset)) === nothing
            end

            # Short vector of bytes in the set
            if !isempty(byteset)
                bytes = rand(collect(byteset), 128)
                @test ScanByte._memchr(T, SizedMemory(bytes), Val(byteset)) == 1
            end
        end
    end

    # Also test gen_scan_function directly (instead of _gen_scan_function)
    byteset = ByteSet([0x02, 0x09, 0x11])
    @test memchr(SizedMemory([1,2,3]), Val(byteset)) == 9
    @test memchr(SizedMemory([1,3,4]), Val(byteset)) === nothing
    @test memchr(SizedMemory([1,5,6,7,4,17,13]), Val(byteset)) == 41

    # Test the memchr method with pointer
    bytes = [1,5,6,7,4,17,13]
    @test memchr(pointer(bytes), UInt(45), Val(byteset)) === 41
    @test memchr(pointer(bytes), UInt(33), Val(byteset)) === nothing

    # Test the memchr method with array/string directly
    bytes = [1,5,6,7,4,17,13]
    @test memchr(UInt8.(bytes), Val(byteset)) === 6
    @test memchr(String(UInt8.(bytes)), Val(byteset)) === 6
    @test memchr("\x03\x02", Val(byteset)) === 2
end


src/byteset.jl
struct ByteSet <: AbstractSet{UInt8}
    data::NTuple{4, UInt64}
    ByteSet(x::NTuple{4, UInt64}) = new(x)
end

ByteSet() = ByteSet((UInt64(0), UInt64(0), UInt64(0), UInt64(0)))
Base.length(s::ByteSet) = mapreduce(count_ones, +, s.data)
Base.isempty(s::ByteSet) = s === ByteSet()

function ByteSet(it)
    a = b = c = d = UInt64(0)
    for i in it
        vi = convert(UInt8, i)
        if vi < 0x40
            a |= UInt(1) << ((vi - 0x00) & 0x3f)
        elseif vi < 0x80
            b |= UInt(1) << ((vi - 0x40) & 0x3f)
        elseif vi < 0xc0
            c |= UInt(1) << ((vi - 0x80) & 0x3f)
        else
            d |= UInt(1) << ((vi - 0xc0) & 0x3f)
        end
    end
    ByteSet((a, b, c, d))
end

function Base.minimum(s::ByteSet)
    y = iterate(s)
    y === nothing ? Base._empty_reduce_error() : first(y)
end

function Base.maximum(s::ByteSet)
    offset = 0x03 * UInt8(64)
    for i in 0:3
        @inbounds bits = s.data[4 - i]
        iszero(bits) && continue
        return ((3-i)*64 + (64 - leading_zeros(bits)) - 1) % UInt8
    end
    Base._empty_reduce_error()
end

function Base.in(byte::UInt8, s::ByteSet)
    i, o = divrem(byte, UInt8(64))
    @inbounds !(iszero(s.data[i & 0x03 + 0x01] >>> (o & 0x3f) & UInt(1)))
end

@inline function Base.iterate(s::ByteSet, state=UInt(0))
    ioffset, offset = divrem(state, UInt(64))
    n = UInt(0)
    while iszero(n)
        ioffset > 3 && return nothing
        n = s.data[ioffset + 1] >>> offset
        offset *= !iszero(n)
        ioffset += 1
    end
    tz = trailing_zeros(n)
    result = (64 * (ioffset - 1) + offset + tz) % UInt8
    (result, UInt(result) + UInt(1))
end

function Base.:~(s::ByteSet)
    a, b, c, d = s.data
    ByteSet((~a, ~b, ~c, ~d))
end

is_contiguous(s::ByteSet) = isempty(s) || (maximum(s) - minimum(s) + 1 == length(s))

src/codegen.jl
# The overall goal of the code in this file is to create code which can
# use SIMD to find the first byte in a bytevector not contained in a ByteSet.

# Due to historical reasons, this code finds the first byte that is NOT part
# of a byteset. To make it work, I simply invert the input byteset in the
# _gen_function_content method.

# The code proceeds in the following steps in a loop:
# * load 16 (SSSE3) or 32 (AVX2) bytes at a time to a SIMD vector V
# * Use zerovec_* function to zero out all bytes which are in the byteset
# * Use haszerolayout to check if all bytes are zero. If so, skip 16/32 bytes.
# * Else, use leading_zero_bytes to skip that amount ahead.

const v256 = SIMD.Vec{32, UInt8}
const v128 = SIMD.Vec{16, UInt8}
const BVec = Union{v128, v256}
const _ZERO_v256 = v256(ntuple(i -> VecElement{UInt8}(0x00), 32))

# Discover if the system CPU has SSSE or AVX2 instruction sets
let
    llvmpath = if VERSION ≥ v"1.6.0-DEV.1429"
        Base.libllvm_path()
    else
        only(filter(lib->occursin(r"LLVM\b", basename(lib)), Libdl.dllist()))
    end
    libllvm = Libdl.dlopen(llvmpath)
    gethostcpufeatures = Libdl.dlsym(libllvm, :LLVMGetHostCPUFeatures)
    features_cstring = ccall(gethostcpufeatures, Cstring, ())
    features = split(unsafe_string(features_cstring), ',')
    Libc.free(features_cstring)

    # Need both SSE2 and SSSE3 to process v128 vectors.
    @eval const SSSE3 = $(any(isequal("+ssse3"), features) & any(isequal("+sse2"), features))
    @eval const AVX2 = $(any(isequal("+avx2"), features))

    # Prefer 32-byte vectors because larger vectors = higher speed
    @eval const DEFVEC = if get(ENV, "JULIA_CPU_TARGET", "native") != "native"
        nothing
    elseif AVX2
        v256
    elseif SSSE3
        v128
    else
        nothing
    end
end

"""
    vpcmpeqb(a::BVec, b::BVec) -> BVec

Compare vectors `a` and `b` element wise and return a vector with `0x00`
where elements are not equal, and `0xff` where they are. Maps to the `vpcmpeqb`
AVX2 CPU instruction, or the `pcmpeqb` SSE2 instruction.
"""
function vpcmpeqb end

"""
    vpshufb(a::BVec, b::BVec) -> BVec

Maps to the AVX2 `vpshufb` instruction or the SSSE3 `pshufb` instruction depending
on the width of the BVec.
"""
function vpshufb end

"""
    vec_uge(a::BVec, b::BVec) -> BVec

Compare vectors `a` and `b` element wise and return a vector with `0xff`
where `a[i] ≥ b[i]``, and `0x00` otherwise. Implemented efficiently for CPUs
with the `vpcmpeqb` and `vpmaxub` instructions.

See also: [`vpcmpeqb`](@ref)
"""
function vec_uge end

# In this statement, define some functions for either 16-byte or 32-byte vectors
let
    # icmp eq instruction yields bool (i1) values. We extend with sext to 0x00/0xff.
    # since that's the native output of vcmpeqb instruction, LLVM will optimize it
    # to just that.
    vpcmpeqb_template = """%res = icmp eq <N x i8> %0, %1
    %resb = sext <N x i1> %res to <N x i8>
    ret <N x i8> %resb
    """

    uge_template = """%res = icmp uge <N x i8> %0, %1
    %resb = sext <N x i1> %res to <N x i8>
    ret <N x i8> %resb
    """

    for N in (16, 32)
        T = NTuple{N, VecElement{UInt8}}
        ST = SIMD.Vec{N, UInt8}
        instruction_set = N == 16 ? "ssse3" : "avx2"
        instruction_tail = N == 16 ? ".128" : ""
        intrinsic = "llvm.x86.$(instruction_set).pshuf.b$(instruction_tail)"
        vpcmpeqb_code = replace(vpcmpeqb_template, "<N x" => "<$(sizeof(T)) x")

        @eval @inline function vpcmpeqb(a::$ST, b::$ST)
            $(ST)(Base.llvmcall($vpcmpeqb_code, $T, Tuple{$T, $T}, a.data, b.data))
        end

        @eval @inline function vpshufb(a::$ST, b::$ST)
            $(ST)(ccall($intrinsic, llvmcall, $T, ($T, $T), a.data, b.data))
        end

        @eval const $(Symbol("_SHIFT", string(8N))) = $(ST)(ntuple(i -> 0x01 << ((i-1)%8), $N))
        @eval @inline bitshift_ones(shift::$ST) = vpshufb($(Symbol("_SHIFT", string(8N))), shift)

        uge_code = replace(uge_template, "<N x" => "<$(sizeof(T)) x")
        @eval @inline function vec_uge(a::$ST, b::$ST)
            $(ST)(Base.llvmcall($uge_code, $T, Tuple{$T, $T}, a.data, b.data))
        end
    end
end

"""
    haszerolayout(x::BVec) -> Bool

Test if the vector consists of all zeros.
"""
@inline haszerolayout(v::BVec) = iszero(mapreduce(i -> i.value, |, v.data))

"Count the number of 0x00 bytes in a vector"
@inline function leading_zero_bytes(v::v256)
    # First compare to zero to get vector of 0xff where is zero, else 0x00
    # Then use vpcmpeqb to extract top bits of each byte to a single UInt32,
    # which is a bitvector, where the 1's were 0x00 in the original vector
    # Then use trailing/leading ones to count the number
    eqzero = vpcmpeqb(v, _ZERO_v256).data
    packed = ccall("llvm.x86.avx2.pmovmskb", llvmcall, UInt32, (NTuple{32, VecElement{UInt8}},), eqzero)
    @static if ENDIAN_BOM == 0x04030201
        return trailing_ones(packed)
    else
        return leading_ones(packed)
    end
end

# vpmovmskb requires AVX2, so we fall back to this.
@inline function leading_zero_bytes(v::v128)
    lzbits = @static if ENDIAN_BOM == 0x04030201
        trailing_zeros(reinterpret(UInt128, v))
    else
        leading_zeros(reinterpret(UInt128, v))
    end
    div(lzbits % UInt, 8)
end

@inline loadvector(::Type{T}, p::Ptr) where {T <: BVec} = unsafe_load(Ptr{T}(p))

# We have this as a separate function to keep the same constant mask in memory.
@inline shrl4(x) = x >>> 0x04

### ---- ZEROVEC_ FUNCTIONS
# The zerovec functions takes a single vector x of type BVec, and some more arguments
# which are supposed to be constant folded. The constant folded arguments are computed
# using a ByteSet. The resulting zerovec code will highly efficiently zero out the
# bytes which are contained in the original byteset.

# This is the generic fallback. For each of the 16 possible values of the lower 4
# bytes, we use vpshufb to get a byte B. That byte encodes which of the 8 values of
# H = bits 5-7 of the input bytes are not allowed. Let's say the byte is 0b11010011.
# Then values 3, 4, 6 are allowed. So we compute if B & (0x01 << H), that will be
# zero only if the byte is allowed.
# For the highest 8th bit, we exploit the fact that if that bit is set, vpshufb
# always returns 0x00. So either `upper` or `lower` will be 0x00, and we can or it
# together to get a combined table.
# See also http://0x80.pl/articles/simd-byte-lookup.html for explanation.
@inline function zerovec_generic(x::T, topzero::T, topone::T) where {T <: BVec}
    lower = vpshufb(topzero, x)
    upper = vpshufb(topone, x ⊻ 0b10000000)
    bitmap = lower | upper
    return bitmap & bitshift_ones(shrl4(x))
end

# Essentially like above, except in the special case of all accepted values being
# within 128 of each other, we can shift the acceptable values down in 0x00:0x7f,
# and then only use one table. if f == ~, and input top bit is set, the bitmap
# will be 0xff, and it will fail no matter the shift.
# By changing the offset and/or setting f == identity, this function can also cover
# cases where all REJECTED values are within 128 of each other.
@inline function zerovec_128(x::T, lut::T, offset::UInt8, f::Function) where {T <: BVec}
    y = x - offset
    bitmap = f(vpshufb(lut, y))
    return bitmap & bitshift_ones(shrl4(y))
end

# If there are only 8 accepted elements, we use a vpshufb to get the bitmask
# of the accepted top 4 bits directly.
@inline function zerovec_8elem(x::T, lut1::T, lut2::T) where {T <: BVec}
    # Get a 8-bit bitarray of the possible ones
    mask = vpshufb(lut1, x & 0b00001111)
    shifted = vpshufb(lut2, shrl4(x))
    return vpcmpeqb(shifted, mask & shifted)
end

# One where it's a single range. After subtracting low, all accepted values end
# up in 0x00:len-1, and so a >= (aka. uge) check will zero out accepted bytes.
@inline function zerovec_range(x::BVec, low::UInt8, len::UInt8)
    vec_uge((x - low), typeof(x)(len))
end

# One where, in all the disallowed values, the lower nibble is unique.
# This one is surprisingly common and very efficient.
# If all 0x80:0xff are allowed, the mask can be 0xff, and is compiled away
@inline function zerovec_inv_nibble(x::T, lut::T, mask::UInt8) where {T <: BVec}
    # If upper bit is set, vpshufb yields 0x00. 0x00 is not equal to any bytes with the
    # upper biset set, so the comparison will return 0x00, allowing it.
    return vpcmpeqb(x, vpshufb(lut, x & mask))
end


# Same as above, but inverted. Even better!
@inline function zerovec_nibble(x::T, lut::T, mask::UInt8) where {T <: BVec}
    return x ⊻ vpshufb(lut, x & mask)
end

# Simplest of all - and fastest!
@inline zerovec_same(x::BVec, y::UInt8) = x ⊻ y

function load_lut(::Type{T}, v::Vector{UInt8}) where {T <: BVec}
    T === v256 && (v = repeat(v, 2))
    return unsafe_load(Ptr{T}(pointer(v)))
end  

# Compute the table (aka. LUT) used to generate the bitmap in zerovec_generic.
function generic_luts(::Type{T}, ::Val{byteset}, offset::UInt8, invert::Bool) where {
    byteset, T <: BVec}
    # If ascii, we set each allowed bit, but invert after vpshufb. Hence, if top bit
    # is set, it returns 0x00 and is inverted to 0xff, guaranteeing failure
    topzero = fill(invert ? 0xff : 0x00, 16)
    topone = copy(topzero)
    for byte in byteset
        byte -= offset
        # Lower 4 bits is used in vpshufb, so it's the index into the LUT
        index = (byte & 0x0f) + 0x01
        # Upper bit sets which of the two bitmaps we use.
        bitmap = (byte & 0x80) == 0x80 ? topone : topzero
        # Bits 5,6,7 from lowest control the shift. If, after a shift, the bit
        # aligns with a zero, it's in the bitmask
        shift = (byte >> 0x04) & 0x07
        bitmap[index] ⊻= 0x01 << shift
    end
    return load_lut(T, topzero), load_lut(T, topone)
end

# Compute the LUT for use in zerovec_8elem.
function elem8_luts(::Type{T}, ::Val{byteset}) where {byteset, T <: BVec}
    allowed_mask = fill(0xff, 16)
    bitindices = fill(0x00, 16)
    for (i, byte) in enumerate(byteset)
        bitindex = 0x01 << (i - 1)
        allowed_mask[(byte & 0x0f) + 0x01] ⊻= bitindex
        bitindices[(byte >>> 0x04) + 0x01] ⊻= bitindex
    end
    return load_lut(T, allowed_mask), load_lut(T, bitindices)
end

# Compute LUT for zerovec_unique_nibble.
function unique_lut(::Type{T}, ::Val{byteset}, invert::Bool) where {byteset, T <: BVec}
    # The default, unset value of the vector v must be one where v[x & 0x0f + 1] ⊻ x
    # is never accidentally zero.
    allowed = collect(0x01:0x10)
    for byte in (invert ? ~byteset : byteset)
        allowed[(byte & 0b00001111) + 1] = byte
    end
    return load_lut(T, allowed)
end 

### ---- GEN_ZERO_ FUNCTIONS
# These will take a symbol sym and a byteset x. They will then compute all the
# relevant values for use in the zerovec_* functions, and produce code that
# will zero out the accepted bytes. Notably, the calculated values will all
# be compile-time constants, EXCEPT the symbol, which represents the input vector.
@inline @generated function zero_generic(x::BVec, ::Val{bs}) where bs
    lut1, lut2 = generic_luts(x, Val{bs}(), 0x00, true)
    :(zerovec_generic(x, $lut1, $lut2))
end

@inline @generated function zero_8elem(x::BVec, ::Val{bs}) where bs
    lut1, lut2 = elem8_luts(x, Val{bs}())
    :(zerovec_8elem(x, $lut1, $lut2))
end

@inline @generated function zero_128(x::BVec, ::Val{bs}, ::Val{ascii}, ::Val{inverted}) where {
    bs, ascii, inverted
}
    if ascii && !inverted
        offset, f, invert = 0x00, ~, false
    elseif ascii && inverted
        offset, f, invert = 0x80, ~, false
    elseif !ascii && !inverted
        offset, f, invert = minimum(bs), ~, false
    else
        offset, f, invert = minimum(~bs), identity, true
    end
    (lut, _) = generic_luts(x, Val{bs}(), offset, invert)
    :(zerovec_128(x, $lut, $offset, $f))
end

@inline @generated function zero_range(x::BVec, ::Val{bs}) where bs
    :(zerovec_range(x, $(minimum(bs)), $(UInt8(length(bs)))))
end

@inline @generated function zero_inv_range(x::BVec, ::Val{bs}) where bs
    # An inverted range is the same as a shifted range, because UInt8 arithmetic
    # is circular. So we can simply adjust the shift, and return regular vec_range
    :(zerovec_range(x, $(maximum(~bs) + 0x01), $(UInt8(length(bs)))))
end

@inline @generated function zero_nibble(x::BVec, ::Val{bs}, ::Val{invert}) where {bs, invert}
    lut = unique_lut(x, Val{bs}(), invert)
    mask = maximum(invert ? ~bs : bs) > 0x7f ? 0x0f : 0xff
    if invert
        :(zerovec_inv_nibble(x, $lut, $mask))
    else
        :(zerovec_nibble(x, $lut, $mask))
    end
end

@inline @generated function zero_same(x::BVec, ::Val{bs}) where bs
    :(zerovec_same(x, $(minimum(bs))))
end

### ----- GEN ZERO CODE
# This is the main function of this file. Given a BVec type T and a byteset B,
# it will produce the most optimal code which will zero out all bytes in input
# vector of type T, which are present in B. This function tests for the most
# efficient special cases, in order, until finally defaulting to generics.
@inline @generated function zero_code(x::BVec, valbs::Val{bs}) where bs
    # These special cases are handled completely differently
    # and dispatch should not reach here if this is the case
    @assert !in(length(bs), (0, 255, 256))

    if length(bs) == 1
        :(zero_same(x, valbs))
    elseif length(bs) == length(Set([i & 0x0f for i in bs]))
        :(zero_nibble(x, valbs, Val(false)))
    elseif length(~bs) == length(Set([i & 0x0f for i in ~bs]))
        :(zero_nibble(x, valbs, Val(true)))
    elseif is_contiguous(bs)
        :(zero_range(x, valbs))
    elseif is_contiguous(~bs)
        :(zero_inv_range(x, valbs))
    elseif minimum(bs) > 127
        :(zero_128(x, valbs, $(Val(true)), $(Val(true))))
    elseif maximum(bs) < 128
        :(zero_128(x, valbs, $(Val(true)), $(Val(false))))
    elseif maximum(~bs) - minimum(~bs) < 128
        :(zero_128(x, valbs, $(Val(false)), $(Val(true))))
    elseif maximum(bs) - minimum(bs) < 128
        :(zero_128(x, valbs, $(Val(false)), $(Val(false))))
    elseif length(bs) < 9
        :(zero_8elem(x, valbs))
    else
        :(zero_generic(x, valbs))
    end
end

"""
    SizedMemory

This is an internal type, and is considered unstable.
Construct a `SizedMemory` from a string, or something that implements `pointer` and
`sizeof`. This struct simply wraps a pointer and a length and is completely unsafe.
Care must be taken to ensure the underlying memory isn't garbage collected or moved.
"""
struct SizedMemory
    ptr::Ptr{UInt8}
    len::UInt
end

Base.pointer(mem::SizedMemory) = mem.ptr
Base.length(mem::SizedMemory) = mem.len
Base.isempty(mem::SizedMemory) = iszero(length(mem))
SizedMemory(x) = SizedMemory(pointer(x), sizeof(x))

"""
    memchr(x, bytes)
    memchr(ptr::Ptr, len::UInt, bytes)

Return first position of any byte in `bytes`, in memory `mem`, or
`nothing` if no bytes were found

`bytes` can be a `Val{::ByteSet}`, in which case this function specializes 
to the byteset, or a single `UInt8`, in which case it does not.

`x` can be any type that implements `pointer` and `sizeof`, or alternatively
a pointer and a memory length can be passed.
"""
function memchr end

@inline function _memchr_nonempty(::Type{T}, mem::SizedMemory, bs::Val) where {T <: BVec}
    local zeroed
    local nscanned
    nscanned = zero(UInt)
    while true
        vector = loadvector(T, pointer(mem) + nscanned)
        zeroed = zero_code(vector, bs)
        haszerolayout(zeroed) || break
        nscanned += sizeof(T)
        nscanned > length(mem) && break
    end
    nscanned += leading_zero_bytes(zeroed) + 1
    nscanned > length(mem) ? nothing : nscanned % Int
end

@inline function _memchr_nonempty(::Nothing, mem::SizedMemory, valbs::Val{byteset}) where byteset
    for i in Base.OneTo(mem.len)
        in(unsafe_load(mem.ptr + i - 1), byteset) || return i % Int
    end
    nothing
end

@inline @generated function _memchr(T, mem::SizedMemory, ::Val{byteset}) where byteset
    byteset isa ByteSet || error("memchr must be passed a Val of a ByteSet instance")
    len = length(byteset)
    # Three special cases:
    # 1) Empty byteset: The byte is never found and always returns nothing
    # 2) Byteset is all bytes: Byte is found at pos 1 iff mem is not empty
    # 3) One-element byteset: Just use the non-byteset memchr method
    if iszero(len)
        quote nothing end
    elseif isone(len)
        return :(memchr(mem, $(first(byteset))))
    elseif len == 256
        :(ifelse(isempty(mem), nothing, 1))
    else
        # We invert the byteset, because for historical reasons, this package
        # was originally written to find the first byte NOT in a byteset
        inverted = ~byteset
        :(_memchr_nonempty(T, mem, Val{$(inverted)}()))
    end
end

# Default method given a byteset, dispatches to default vector type
@inline memchr(mem::SizedMemory, valbs::Val) = _memchr(DEFVEC, mem, valbs)

# TODO: We might eventually replace this with a pure Julia function, but
# for now, libc's memchr is 2x faster.
@inline function memchr(mem::SizedMemory, byte::UInt8)
    pos = @ccall memchr(pointer(mem)::Ptr{UInt8}, byte::Cint, length(mem)::Csize_t)::Ptr{Cchar}
    pos == C_NULL ? nothing : ((pos - pointer(mem)) + 1) % Int
end

@inline function memchr(ptr::Ptr, len::UInt, byte_s::Union{UInt8, Val})
    memchr(SizedMemory(Ptr{UInt8}(ptr), len), byte_s)
end

@inline function memchr(x, byte_s::Union{UInt8, Val})
    GC.@preserve x memchr(SizedMemory(x), byte_s)
end
